Simplenote/Classes/SPAuthViewController.swift

```diff
@@ -287,17 +303,50 @@ private extension SPAuthViewController {
         navigationController?.setNavigationBarHidden(false, animated: true)
     }
 
-    func refreshActions() {
-        secondaryActionButton.isHidden = mode.secondaryActionSelector == nil
-        tertiaryActionButton.isHidden = mode.tertiaryActionSelector == nil
+    func refreshInputViews() {
+        let inputElements = mode.inputElements
+        
+        emailInputView.isHidden     = !inputElements.contains(.username)
+        passwordInputView.isHidden  = !inputElements.contains(.password)
+        codeInputView.isHidden      = !inputElements.contains(.code)
+        actionsSeparator.isHidden   = !inputElements.contains(.actionSeparator)
     }
+    
+    func refreshActionViews() {
+        let viewMap: [AuthenticationActionName: UIButton] = [
+            .primary: primaryActionButton,
+            .secondary: secondaryActionButton,
+            .tertiary: tertiaryActionButton,
+            .quaternary: quaternaryActionButton
+        ]
+        
+        for actionView in allActionViews {
+            actionView.isHidden = true
+        }
 
-    func refreshInputViews() {
+        for descriptor in mode.actions {
+            guard let actionView = viewMap[descriptor.name] else {
+                assertionFailure()
+                continue
+            }
+
+            if let title = descriptor.text {
+                actionView.setTitle(title, for: .normal)
+            }
+            
+            if let attributedTitle = descriptor.attributedText {
+                actionView.setAttributedTitle(attributedTitle, for: .normal)
+            }
+
+            actionView.addTarget(self, action: descriptor.selector, for: .touchUpInside)
+            actionView.isHidden = false
+        }
```

Comment:
**Ideally** we don't have 4x Actions, and use them accordingly. Instead, we should simply instantiate the `UIButton` instances, based on the descriptor.

I'm keeping the changeset delta on check, though, and following along. For now!